src/GeneralizedTCRMapping.ts

```diff
@@ -26,46 +26,93 @@ function hexStringToLowerCase(input: string): string {
+    log.warning('GTCR: No address found for itemID {}.', [event.params._itemID.toHexString()]);
     return // Invalid submission. No Op
   }
-  const fpmmAddress = decodedData.slice(addressStartIndex, addressStartIndex + 42)
+  let fpmmAddress = decodedData.slice(addressStartIndex, addressStartIndex + 42);
 
   // Workaround missing String.toLowerCase function in assemblyscript.
-  const lowerCaseFpmmAddr = hexStringToLowerCase(fpmmAddress)
+  let lowerCaseFpmmAddr = hexStringToLowerCase(fpmmAddress);
 
   let fpmm = FixedProductMarketMaker.load(lowerCaseFpmmAddr);
   if (fpmm == null) {
-    log.warning("GTCR: Could not load FPMM for {}", [lowerCaseFpmmAddr]);
+    log.warning("GTCR: Could not load FPMM for {}.", [lowerCaseFpmmAddr]);
     return;
   }
 
-  // Items on a TCR can be in 1 of 4 states:
-  // - (0) Absent: The item is not registered on the TCR and there are no pending requests.
-  // - (1) Registered: The item is registered and there are no pending requests.
-  // - (2) Registration Requested: The item is not registered on the TCR, but there is a pending
-  //       registration request.
-  // - (3) Removal Requested: The item is registered on the TCR, but there is a pending removal
-  //       request.
-  //
-  // Registration and Removal requests can be challenged. Once the request resolves (either by
-  // passing the challenge period or via dispute resolution), the item state is updated to 0 or 1.
-
-  const REGISTERED = 1;
-  const REMOVAL_REQUESTED = 3;
-
-  fpmm.klerosTCRitemID = event.params._itemID.toHexString();
-  fpmm.klerosTCRstatus = itemInfo.value1;
-  fpmm.klerosTCRregistered = fpmm.klerosTCRstatus == REGISTERED || fpmm.klerosTCRstatus == REMOVAL_REQUESTED;
+  // Check if we are updating an entry or creating a new one.
+  let newSubmission = true;
+  for (let i = 0; i < fpmm.submissionIDs.length; i++) {
+    let submissionIDs = fpmm.submissionIDs;
+    let itemID = submissionIDs[i];
+    let submission = KlerosSubmission.load(itemID);
+    if (submission != null && submission.id == event.params._itemID.toHexString()) {
+      newSubmission = false;
+    }
+  }
+
+  let currentSubmission: KlerosSubmission | null = null;
+  if (newSubmission) {
+    currentSubmission = new KlerosSubmission(event.params._itemID.toHexString());
+    currentSubmission.listAddress = event.address.toHexString();
+    currentSubmission.market = fpmm.id;
+
+    let submissionIDs = fpmm.submissionIDs;
+    submissionIDs.push(currentSubmission.id);
+
+    fpmm.submissionIDs = submissionIDs;
+  } else {
+    currentSubmission = KlerosSubmission.load(event.params._itemID.toHexString());
+  }
+
+  currentSubmission.status = getStatus(itemInfo.value1);
+  currentSubmission.save();
+
+  // Check if there is at least one accepted submission. If so,
+  // set klerosTCRregistered = true. Set to false otherwise.
+  fpmm.klerosTCRregistered = false;
+  for (let i = 0; i < fpmm.submissionIDs.length; i++) {
+    let submissionIDs = fpmm.submissionIDs;
```

Comment:
Should this be outside the loop as well?

Reply:
Fixed